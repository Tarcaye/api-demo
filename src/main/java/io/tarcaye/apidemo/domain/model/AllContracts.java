package io.tarcaye.apidemo.domain.model;

import java.util.Optional;

public interface AllContracts {
    Optional<Contracts> findFor(String customerId);
}
