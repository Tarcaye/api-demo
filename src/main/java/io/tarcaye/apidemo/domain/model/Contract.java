package io.tarcaye.apidemo.domain.model;

public class Contract {

    public enum Type {
        LOCATION, ACHAT
    }

    private final long id;
    private final Customer customer;
    private final Type type;

    public Contract(Customer customer, long id, Type type) {
        this.id = id;
        this.customer = customer;
        this.type = type;
    }

    public long getId() {
        return id;
    }

    public Type getType() {
        return type;
    }

    public boolean belongsTo(Customer customer) {
        return this.customer.getId().equals(customer.getId());
    }
}
