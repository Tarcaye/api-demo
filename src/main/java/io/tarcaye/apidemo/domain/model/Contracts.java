package io.tarcaye.apidemo.domain.model;

import java.util.List;

public class Contracts {

    private final List<Contract> values;

    public Contracts(List<Contract> values) {
        this.values = values;
    }

    public List<Contract> asList() {
        return values;
    }
}
