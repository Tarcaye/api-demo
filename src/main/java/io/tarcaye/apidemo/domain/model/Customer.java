package io.tarcaye.apidemo.domain.model;

public class Customer {
    private final Long id;

    private Customer(Long id) {
        this.id = id;
    }

    public static Customer of(String id) {
        return new Customer(Long.valueOf(id));
    }

    public Long getId() {
        return id;
    }
}
