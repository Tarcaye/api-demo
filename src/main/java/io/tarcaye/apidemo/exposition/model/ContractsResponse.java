package io.tarcaye.apidemo.exposition.model;

import io.tarcaye.apidemo.domain.model.Contract;
import io.tarcaye.apidemo.domain.model.Contracts;
import lombok.Getter;

import java.util.List;

import static java.util.Collections.unmodifiableList;
import static java.util.stream.Collectors.toList;

public class ContractsResponse {

    private final List<ContractResponse> contracts;

    public ContractsResponse(List<ContractResponse> contracts) {
        this.contracts = contracts;
    }

    public static ContractsResponse of(Contracts contracts) {
        List<ContractResponse> contractResponses = contracts.asList().stream()
                .map(ContractResponse::from)
                .collect(toList());

        return new ContractsResponse(contractResponses);
    }

    public List<ContractResponse> getContracts() {
        return unmodifiableList(contracts);
    }

    @Getter
    public static class ContractResponse {
        private final long id;
        private final Type type;

        public ContractResponse(long id, Type type) {
            this.id = id;
            this.type = type;
        }

        public static ContractResponse from(Contract contract) {
            return new ContractResponse(contract.getId(), Type.from(contract.getType()));
        }


        enum Type {
            ACHAT, LOCATION;

            static Type from(Contract.Type type) {
                return Type.valueOf(type.name());
            }
        }
    }
}
