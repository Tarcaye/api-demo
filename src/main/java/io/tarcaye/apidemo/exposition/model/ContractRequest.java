package io.tarcaye.apidemo.exposition.model;

import io.tarcaye.apidemo.domain.model.Contract;
import io.tarcaye.apidemo.domain.model.Customer;

public class ContractRequest {

    public String id;

    public String type;

    public ContractRequest(String id, String type) {
        this.type = type;
        this.id = id;
    }

    public Contract toCore(String customerId){
        return new Contract(Customer.of(customerId), Long.parseLong(id), Contract.Type.valueOf(type));
    }
}
