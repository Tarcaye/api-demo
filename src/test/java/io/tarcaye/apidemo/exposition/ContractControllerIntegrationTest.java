package io.tarcaye.apidemo.exposition;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.tarcaye.apidemo.ApiDemoApplication;
import io.tarcaye.apidemo.exposition.model.ContractRequest;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import static org.junit.jupiter.api.Assertions.assertEquals;

@AutoConfigureMockMvc
@SpringBootTest(classes = ApiDemoApplication.class)
class ContractControllerIntegrationTest {

    @Autowired
    private MockMvc mockMvc;

    @Test
    void getContracts() throws Exception {
        String uri = "/customers/9001/contracts";
        MvcResult mvcResult = mockMvc.perform(MockMvcRequestBuilders.get(uri)
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();

        int status = mvcResult.getResponse().getStatus();
        assertEquals(200, status);
        String content = mvcResult.getResponse().getContentAsString();
        Approvals.verify(content);
    }

    @Test
    void addContract() throws Exception {
        String uri = "/customers/21/contracts";
        MvcResult mvcResult = mockMvc.perform(MockMvcRequestBuilders.post(uri)
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapToJson(new ContractRequest("12", "ACHAT")))
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();

        int status = mvcResult.getResponse().getStatus();
        assertEquals(200, status);
        String content = mvcResult.getResponse().getContentAsString();
        Approvals.verify(content);
    }

    @Test
    void addAndRetrieveContracts() throws Exception {
        String uri = "/customers/42/contracts";
        mockMvc.perform(MockMvcRequestBuilders.post(uri)
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapToJson(new ContractRequest("12", "ACHAT")))
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();
        mockMvc.perform(MockMvcRequestBuilders.post(uri)
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapToJson(new ContractRequest("14", "ACHAT")))
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();
        mockMvc.perform(MockMvcRequestBuilders.post(uri)
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapToJson(new ContractRequest("156", "ACHAT")))
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();
        mockMvc.perform(MockMvcRequestBuilders.post(uri)
                .contentType(MediaType.APPLICATION_JSON)
                .content(mapToJson(new ContractRequest("178", "LOCATION")))
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();

        MvcResult mvcResult = mockMvc.perform(MockMvcRequestBuilders.get(uri)
                .accept(MediaType.APPLICATION_JSON_VALUE)).andReturn();

        int status = mvcResult.getResponse().getStatus();
        assertEquals(200, status);
        String content = mvcResult.getResponse().getContentAsString();
        Approvals.verify(content);
    }

    protected String mapToJson(Object obj) throws JsonProcessingException {
        ObjectMapper objectMapper = new ObjectMapper();
        return objectMapper.writeValueAsString(obj);
    }
}