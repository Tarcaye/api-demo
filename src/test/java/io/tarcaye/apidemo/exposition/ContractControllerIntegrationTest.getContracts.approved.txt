{"contracts":[{"id":1,"type":"ACHAT"},{"id":2,"type":"ACHAT"}]}